### PR TITLE
Fix potentional deadlock between xdsproxy and Istiod

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/39209.

This is a pretty complex issue; see the comments in the PR and the
original issue for more discussion

I had a really hard time adding a test for this. I tried for a while. What I did manually was to deploy 8 test namespaces at once (this is a large load, XDS is like 5mb each...). Before, it was almost always 5-10 pods stuck. With this, only delta pods got stuck because I hadn't yet fixed delta xds.

### CODE REVIEW

**Comment by howardjohn** [comment]
Put a hold since I need to fix delta still

File: pkg/istio-agent/xds_proxy.go [hunk 1]
```diff
@@ -319,10 +307,29 @@ func (p *XdsProxy) handleStream(downstream adsStream) error {
 		conID:           connectionNumber.Inc(),
 		upstreamError:   make(chan error, 2), // can be produced by recv and send
 		downstreamError: make(chan error, 2), // can be produced by recv and send
-		requestsChan:    make(chan *discovery.DiscoveryRequest, 10),
-		responsesChan:   make(chan *discovery.DiscoveryResponse, 10),
-		stopChan:        make(chan struct{}),
-		downstream:      downstream,
+		// Requests channel is unbounded. The Envoy<->XDS Proxy<->Istiod system produces a natural
+		// looping of Recv and Send. Due to backpressure introduce by gRPC natively (that is, Send() can
```
Comment (ramaraochavali) [inline pkg/istio-agent/xds_proxy.go:311]: Due to backpressure introduce -> Due to backpressure introduced

**Review by GregHanson — APPROVED** [review]

**Review by lei-tang — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
Would prefer https://github.com/istio/istio/pull/39309, have you tested that patch?

File: pkg/channels/unbounded.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+// NewUnbounded returns a new instance of Unbounded.
+func NewUnbounded() *Unbounded {
+	return &Unbounded{c: make(chan interface{}, 1)}
+}
+
+// Put adds t to the unbounded buffer.
+// Put will never block
+func (b *Unbounded) Put(t interface{}) {
+	b.mu.Lock()
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/channels/unbounded.go:54]: fwiw this is copied directly from grpc's code but I agree

**Pull request merged**